### PR TITLE
[Local fork] TC 39 decorators

### PR DESCRIPTION
**Pull request by mweststrate**
Local fork with additional updates to @Matchlighter 's awesome work. 

Follow progress in #3638

### CODE REVIEW

**Comment by mweststrate** [comment]
@render preview

File: packages/mobx/src/types/observableobject.ts [hunk 1]
```diff
@@ -712,6 +712,22 @@ function getCachedObservablePropDescriptor(key) {
     )
 }
 
+const fallthroughDescriptorCache = Object.create(null)
+
+export function getCachedFallthroughPropDescriptor(key) {
```
Comment (urugator) [inline packages/mobx/src/types/observableobject.ts:717]: Is it used anywhere?

Comment (mweststrate) [reply packages/mobx/src/types/observableobject.ts]: Nope, nice catch!

**Comment by williamdes** [comment]
This triggers a compilation error.
6.10.x are the last working versions for me.
Should TypeScript 5.x be required ? I am using TS 4.
```
Error: node_modules/mobx/dist/api/annotation.d.ts:11:42 - error TS2304: Cannot find name 'DecoratorContext'.

11     decorate_20223_(value: any, context: DecoratorContext): any;
                                            ~~~~~~~~~~~~~~~~


Error: node_modules/mobx/dist/api/decorators.d.ts:19:68 - error TS2304: Cannot find name 'DecoratorContext'.

19 export declare function is20223Decorator(context: any): context is DecoratorContext;
                                                                      ~~~~~~~~~~~~~~~~


Error: node_modules/mobx/dist/api/decorators.d.ts:20:59 - error TS2304: Cannot find name 'DecoratorContext'.

20 export declare function assert20223DecoratorType(context: DecoratorContext, types: DecoratorContext["kind"][]): void;
                                                             ~~~~~~~~~~~~~~~~


Error: node_modules/mobx/dist/api/decorators.d.ts:20:84 - error TS2304: Cannot find name 'DecoratorContext'.

** Angular Live Development Server is listening on 0.0.0.0:4200, open your browser on http://localhost:4200/ **


20 export declare function assert20223DecoratorType(context: DecoratorContext, types: DecoratorContext["kind"][]): void;
                                                                                      ~~~~~~~~~~~~~~~~


Error: node_modules/mobx/dist/types/decorator_fills.d.ts:1:71 - error TS2552: Cannot find name 'ClassAccessorDecoratorTarget'. Did you mean 'ClassAccessorDecorator'?

1 export type ClassAccessorDecorator<This = any, Value = any> = (value: ClassAccessorDecoratorTarget<This, Value>, context: ClassAccessorDecoratorContext) => ClassAccessorDecoratorResult<This, Value> | void;
                                                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Error: node_modules/mobx/dist/types/decorator_fills.d.ts:1:123 - error TS2552: Cannot find name 'ClassAccessorDecoratorContext'. Did you mean 'ClassAccessorDecorator'?

1 export type ClassAccessorDecorator<This = any, Value = any> = (value: ClassAccessorDecoratorTarget<This, Value>, context: ClassAccessorDecoratorContext) => ClassAccessorDecoratorResult<This, Value> | void;
                                                                                                                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Error: node_modules/mobx/dist/types/decorator_fills.d.ts:1:157 - error TS2552: Cannot find name 'ClassAccessorDecoratorResult'. Did you mean 'ClassAccessorDecorator'?

1 export type ClassAccessorDecorator<This = any, Value = any> = (value: ClassAccessorDecoratorTarget<This, Value>, context: ClassAccessorDecoratorContext) => ClassAccessorDecoratorResult<This, Value> | void;
                                                                                                                                                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✖ Failed to compile.

Error: node_modules/mobx/dist/types/decorator_fills.d.ts:2:101 - error TS2552: Cannot find name 'ClassGetterDecoratorContext'. Did you mean 'ClassGetterDecorator'?

2 export type ClassGetterDecorator<This = any, Value = any> = (value: (this: This) => Value, context: ClassGetterDecoratorContext) => ((this: This) => Value) | void;
                                                                                                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~


Error: node_modules/mobx/dist/types/decorator_fills.d.ts:3:114 - error TS2552: Cannot find name 'ClassSetterDecoratorContext'. Did you mean 'ClassSetterDecorator'?

3 export type ClassSetterDecorator<This = any, Value = any> = (value: (this: This, value: Value) => void, context: ClassSetterDecoratorContext) => ((this: This, value: Value) => void) | void;
                                                                                                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~


Error: node_modules/mobx/dist/types/decorator_fills.d.ts:4:114 - error TS2552: Cannot find name 'ClassMethodDecoratorContext'. Did you mean 'ClassMethodDecorator'?

4 export type ClassMethodDecorator<This = any, Value extends (...p: any[]) => any = any> = (value: Value, context: ClassMethodDecoratorContext<This, Value>) => Value | void;
                                                                                                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~


Error: node_modules/mobx/dist/types/decorator_fills.d.ts:5:113 - error TS2552: Cannot find name 'ClassFieldDecoratorContext'. Did you mean 'ClassFieldDecorator'?

5 export type ClassFieldDecorator<This = any, Value extends (...p: any[]) => any = any> = (value: Value, context: ClassFieldDecoratorContext<This, Value>) => Value | void;
                                                                                                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by mweststrate** [comment]
Yes please upgrade TS. TS doesn't support semantic versioning so we don't
account for that in our releases.

**Pull request merged**